v2 further work

steelbrain opened the pull request:
Docs checklist
- [ ] Introduction
  - [x] Getting started
  - [ ] CLI Usage
  - [ ] Configuration
  - [x] Components
  - [x] Presets
  - [ ] ...
- [ ] Technical
  - [x] Lifecycles
  - [x] Components
  - [x] Presets
  - [ ] ...

steelbrain committed:
:bug: Pipe errors through reporter in non-watcher mode too

steelbrain committed:
:bug: Allow non-relative paths in output config for CLI

steelbrain committed:
:bug: Improve defaults in CLI and Dev for sourceMapPath

steelbrain committed:
:memo: Add basic root and lifecycle docs

steelbrain committed:
:bug: Fix eslint ignore for CI

steelbrain committed:
:bug: Fix specs for pundle-api to make CI pass

steelbrain committed:
:arrow_up: Bump version to v2beta4

steelbrain committed:
:bug: Show a message when start installing package in npm installer

steelbrain committed:
:bug: Fix sourcemap and root directory typos in cli

steelbrain committed:
:art: Improve default configs

steelbrain committed:
:arrow_up: Bump version to v2beta5

steelbrain committed:
:memo: Add some descriptions to configs

steelbrain committed:
:new: Add API version validation to components

steelbrain committed:
:art: Minor style tweak in pundle-api

steelbrain committed:
:new: Call the lifecycle events with component configuration

steelbrain committed:
:new: Add introduction and partial technical docs for Components

steelbrain committed:
:fire: Remove component type simple

steelbrain committed:
:new: Add config file for direnv

steelbrain committed:
:new: Lay foundations for watcher components

steelbrain committed:
:art: Cleanups and code duplication fixes

steelbrain committed:
:new: Invoke new watcher component callbacks instead of config callbacks

steelbrain committed:
:arrow_up: Upgrade internals to use new component APIs

steelbrain committed:
:minibus: Move lifecycles from introduction to technical

steelbrain committed:
:art: Minor improvements in several places

steelbrain committed:
:memo: Add docs about writing a Loader Component

steelbrain committed:
:memo: Add technical docs for Resolver and Plugin components

steelbrain committed:
:memo: Add technical docs for Reporter, Generator and Transformer

steelbrain committed:
:memo: Add Technical docs for Watcher and Post Transformer components

steelbrain committed:
:art: Typo fixes in Components - Technical

natew commented on docs/technical/components.md:
```diff
@@ -13,9 +13,9 @@ It is important to note that all callback methods of the component are invoked w
 Another important thing to note is that all callback methods receive the user-config merged with default config as their first parameter. The given config object is cloned for every invocation so you cannot use it to store state of your component.
 
-All components regardless of their types have two lifecycle methods, `activate()` and `dispose()`. They are invoked when a component is added to Pundle and when Pundle is disposed or the component is removed respectively.
+All components regardless of their types have atleast two lifecycle methods, `activate()` and `dispose()`. They are invoked when a component is added to Pundle and when Pundle is disposed or the component is removed respectively.
```
missing space between `at least`

natew commented on docs/technical/components.md:
```diff
@@ -117,9 +117,9 @@ module.exports = createPlugin(function(config, file) {
+Reporter Components are used to output the reports of Pundle and it's components. It has a main callback that recieves an error or any `Issue` constructed by `pundle-api` package.
 
-Reporter Components are where all of `pundle.compilation.report(...)` results goto. Each report is fed to all of registered reporter components. Each of them can have different means of output. For example, the default reporter writes the errors and other output to CLI, you could write one to output to websockets etc.
+Reporter Components are where all of `pundle.compilation.report(...)` input goes to. Each report is fed to all of registered reporter components. They can have different means of output. For example, the default reporter writes the errors and other output to CLI, you could write one to output to websockets etc.
```
you can just write `input goes` btw, but `input goes to` works

natew commented on docs/technical/components.md:
```diff
@@ -174,7 +174,7 @@ module.exports = createPostTransformer(function(config, contents) {
 ## Watcher Components
 
-Watcher Components are Pundle components that have more than one main callback. They are used when Pundle is running in watcher or dev server mode. Watcher Components usually have `tick`, `ready` and `compile` callbacks.
+Watcher Components are Pundle components that have more than one main callback. They are used when Pundle is running in watcher or dev server mode. Watcher Components have `tick`, `ready` and `compile` callbacks, all of which are optional; though you must at least specify one of them.
```
In this context parens is more appropriate I think:

```
, all of which are optional (though you must at least specify one of them).
```

steelbrain committed:
:memo: Add docs about Presets

steelbrain committed:
:new: Add smarter default for debug config

steelbrain committed:
:new: Add HMR debugging support

steelbrain committed:
:bug: Fix a race condition in HMR server when multiple files are changed at once

steelbrain committed:
:bug: Fix duplicate entries in files changed really quickly in HMR server

steelbrain committed:
:art: Switch between uws and ws seemlessly

steelbrain committed:
:new: Improve pundle init

steelbrain committed:
:new: Add config.output.rootDirectory

steelbrain committed:
:new: Handle the case when no reporter is configured

steelbrain committed:
:new: Tell the user when no resolver is configured

steelbrain committed:
:art: Replace erorrs with MessageIssues during config validation

steelbrain committed:
:art: Improve default reporter output for pundle-api Messages

natew committed:
:new: logo

natew committed:
:new: add fancy readme logo

natew committed:
:bug: fix path for logo

steelbrain committed:
:bug: Fix logo url

natew committed:
:new: words logo

natew committed:
:new: move to words

steelbrain committed:
:bug: Transform typescript even in node_modules if extension matches

steelbrain committed:
:fire: Minor cleanups in resolver

steelbrain committed:
:bug: Fix watcher issues by moving away from chokidar
We're now using watchr that doesn't show this behavior. God. This. was. hard. to. investigate.

steelbrain committed:
:art: Minor ordering fix

steelbrain committed:
:bug: Fix parent requires of children in HMR

steelbrain committed:
:fire: Remove duplicates check in HMR
We now support even direct-children requiring their parents indefinately in HMR

steelbrain committed:
:bug: Reset module invoked status after HMR

steelbrain committed:
:new: Include transformers in allKnownExtensions

steelbrain committed:
:fire: Minor cleanup in wrappers

steelbrain committed:
:new: Try after 10 seconds if server is down, then die

steelbrain committed:
:new: Show full error stacks if debug env var is present

steelbrain committed:
:new: Add stub module for fs

steelbrain committed:
:minibus: Switch to slow babel-traverse from our fast traverse
Our traverse implementation did not allow us to know if a variable had been declared in the parent scope. We needed this to work with dist bundle files of modules like mapbox-gl, because they are already bundled and have their very own require defined.
In our previous implementation, we didn't know they had defined their own require, or on which scope, and because they were not using the variable require directly but instead assigning it to another variable and then using that variable, it made it pretty hard to predict. Therefore I've decided to use the slower but more verbose babel-traverse implementation, now we can now specifically where the require is defined, and only not process requests of that require

steelbrain committed:
:art: Make CI happy

steelbrain committed:
:fire: Remove an unrequired fix

steelbrain committed:
:arrow_down: Revert back to chokidar
other watcher doesn't have the bug we were trying to escape, but it also is quite heavy on cpu during init and not as reliable in general

steelbrain committed:
:art: Minor code arrangement change

steelbrain committed:
:new: Detect re-added files

steelbrain committed:
:new: Gracefully recover from renames, deletions and other errors

steelbrain committed:
:art: Ditch external fs in favor of pundle-fs

steelbrain committed:
:new: Improve default packageMains

steelbrain committed:
Merge branch 'master' into steelbrain/docs

steelbrain committed:
:bug: Fix windows support

steelbrain committed:
:fire: Remove linker

steelbrain committed:
:bug: Fix the params passed to ready() callback of watcher

steelbrain committed:
:arrow_up: Bump version to 2.0.0-beta6

steelbrain committed:
:bug: Remove module main for now as browsers don't natively support it

steelbrain committed:
:arrow_up: Bump version to v2b7

steelbrain committed:
:bug: More stability fixes for the new error recovery stuff

steelbrain committed:
:new: Manually add a uncaughtException handler

steelbrain committed:
:arrow_up: Bump verison to v2b8

steelbrain committed:
:new: Readd modules in typescript preset

steelbrain committed:
:art: Minor usability improvements in cli

steelbrain committed:
:new: Use reporter-cli even when pundle failed on init

steelbrain committed:
:new: Show fancy error message when a module is not installed on root

steelbrain committed:
:new: Add a default manifest for new packages

steelbrain committed:
:art: Accept HMR in example app

steelbrain committed:
:memo: Add getting started docs

steelbrain committed:
:art: Improve logs of dev package

steelbrain committed:
:bug: Fix a prioritization issue

steelbrain committed:
:new: Handle renames smartly without breaking

steelbrain committed:
:new: Include source and target manifest information from resolver

steelbrain committed:
:new: Add plugin-dedupe :tada:

steelbrain committed:
:art: Misc fixes around the codebase

steelbrain committed:
:arrow_up: Bump version to v2beta9

steelbrain committed:
:memo: Mention new dedupe plugin in default configs

steelbrain committed:
:bug: Fix getModuleName() for scoped modules

steelbrain committed:
:arrow_up: Bump version to v2b10

steelbrain committed:
:new: Add unhandledRejection handler in CLI

steelbrain committed:
:bug: Fix 100% cpu usage of pundle after changing dozens of files at once

steelbrain committed:
:new: In dedupe, always use the highest available version

steelbrain committed:
:bug: Fix empty HMR when batch updating files

steelbrain committed:
:new: Add ENV variable to allow dedupe debugging

steelbrain committed:
:bug: Followup fixes

steelbrain committed:
:fire: Remove require.resolve from examples

steelbrain committed:
:new: Use timestamp as sourceMapNamespace for HMR

steelbrain committed:
:art: Minor helpers refactor

steelbrain committed:
:new: Increase imports check debounce time to 100ms

steelbrain committed:
:bug: Fix normal output wrapper

steelbrain committed:
:arrow_up: Bump version to v2b11

steelbrain committed:
:bug: Fix invalid automatic value of config.debug

steelbrain committed:
:bug: Fix default of config.server.sourceMap

steelbrain committed:
:arrow_up: Bump version to v2b12

steelbrain committed:
:new: Use fixed versions

steelbrain committed:
:new: Handle SyntaxErrors in reporter properly

steelbrain committed:
:new: Update compile script

steelbrain committed:
:new: Sort all files by filePath in generator

steelbrain committed:
:racehorse: Make HMR efficient again
We are no longer generating full on each HMR but we're generating them on each HTTP request

One known bug as of this release is double HMR, I'm investigating it

steelbrain committed:
:new: Pass on legit errors from resolver like EMFILE

steelbrain committed:
:fire: Remove support for pluggable filesystem because useless

steelbrain committed:
:new: Use graceful-fs

natew committed:
:bug: fix zindex of errors

steelbrain committed:
:new: Standardize Resolved

steelbrain committed:
:bug: Fix regressions from last commit

steelbrain committed:
:bug: Fix race condition in dev server

steelbrain committed:
:arrow_up: Upgrade to the latest version of Babel

steelbrain committed:
:bug: Add a missing dep

steelbrain committed:
:bug: Do not npm ignore src for pundle-api
Fixes #118

steelbrain committed:
Revert ":arrow_up: Upgrade to the latest version of Babel"

This reverts commit 82abab05d091469f9b58787e56b11858d3840151.

steelbrain committed:
:arrow_up: Bump version to v2b13

steelbrain committed:
:new: Add lastModified to File type

steelbrain committed:
:new: Add helpers about uniqueID to compilation
these are helpful for serialization/deserializing state of Pundle. They will help us when we want to boot from cache

steelbrain committed:
:new: Support resuming watch from an existing files map

steelbrain committed:
:bug: Fix a bug in triggerDebouncedImportsCheck

steelbrain committed:
:art: Fix a flow warning

steelbrain committed:
:bug: Fix a recent regression

steelbrain committed:
:new: Make boot from cache work!

steelbrain committed:
:arrow_up: Bump sb-config-file to v2.0.0

steelbrain committed:
:arrow_up: Upgrade to latest async version of sb-config-file

steelbrain committed:
:memo: Show an info message when restoring from cache

steelbrain committed:
:bug: Upgrade sb-event-kit to fix error

steelbrain committed:
:racehorse: Only write to state file before dispose

steelbrain committed:
:bug: Fix a typo

steelbrain commented:
I wish we could keep this PR focused but Pundle is still in development so this PR is not so much focused as docs aren't the main priority right now, merging as is, will still keep releasing betas from master
